### ISSUE TITLE
nw.Window.open callback returns undefined object when using new_instance

### ISSUE THREAD

**Issue by tolrtd** [nwjs/nw.js#4418]
In nwjs 0.13-beta6, win64 SDK build, when opening a new window with "new_instance" set to true, the Window.open callback returns an undefined object.  Even the 'this' object is undefined.

```
nw.Window.open('new.html', {
   'new_instance': true
},function(win){
   // win is undefined here
});
```

Edit: Unless this is by design because the new window will have a new context that we are not supposed to access?

**Comment by rogerwang** [comment]
This is currently a limitation, will fix it. Before that you can call nw.Window.get() in the new window ...

**Comment by tolrtd** [comment]
I'm trying to get the handle of the new_instance child window from the parent window so that I can close it from the parent.  What would you suggest as the proper way to do this right now, since the nw.* contexts are not shared?

**Comment by tolrtd** [comment]
To add to this - it seems that when I use new_instance = true on any nw.Window.open, all future windows that open have this issue with the callback object being undefined, even if I explicitly set new_instance = false.

**Comment by FiatPax** [comment]
+1 for this issue to be fixed.

**Comment by tolrtd** [comment]
Roger, any idea when this will be fixed?

**Comment by rogerwang** [comment]
I'll see if it can catch the 0.13.0 release.

**Comment by Gimmeaphatbeat** [comment]
+1. `loaded` event also not firing

**Comment by vankasteelj** [comment]
Still happening in 0.13.2

**Comment by kit92** [comment]
also despite a variety of implementations, in v14.1 sdk, calling nw.Window.get() only gets the first window that was created, not the focused window.

**Comment by kailniris** [comment]
after creating new window with
nw.Window.open('new.html', {
   'new_instance': true
},function(win){
   // win is undefined here
});

in the new window nw object will be undefined (e.g: in thew new window console)

**Comment by alexsmith540** [comment]
I fought this same issue most of today as well and finally got it to work (v0.15, windows 10 box).
If you're opening an external URL like I am in a new_instance, i.e. nw.Window.open('https://bar.foo.com',... you need to do a couple of things to make nw.Window.open callback return your Window.
- inside your package.json manifest, set "node-remote":"https://*.foo.com" (you can pass in an array of string patterns too)
- If the content displayed uses requirejs or any other require that might get clobbered my mixed context mode, you absolutely MUST NOT run your app with --mixed-context

For some wild reason that all made my new_instance window.open callbacks start returning the Window and not undefined. 
#voodoomagic

**Comment by kailniris** [comment]
@rogerwang 
I tried out new_instance again in 1.7.0 version, the nw.Window.open callback still return undefined but in the new window the nw object is no longer undefined.
Its pretty handy, the new window is on another thread i guess cuz eg. window.print() does not freeze the main window. The only thing is needed is access the new window context, can it even be achieved somehow considering the thing the other window is on other thread?

**Comment by edi9999** [comment]
Same issue for me, it would be great to fix this :-)

**Comment by edi9999** [comment]
From what I can see, it seems that this doesn't work since the 0.13 migration (eg I tried on 0.14.7 and it doesn't seem to work either).

**Comment by freeneutron** [comment]
Doesn't work in v0.18.0

**Comment by gpetrov** [comment]
still a bug in  0.18.5 ... hope it will be fixed soon, we really need it

**Comment by vankasteelj** [comment]
We can handle some things with the `chrome` api, but we don't have access to DOM or anything truly usefull :/

**Comment by gpetrov** [comment]
well having a new window as nwjs instance gives us a lot of freedom to separate heavy duty tasks that normally block the UI, while still having the full power of web and node together.

Also this gives us a great level of isolation - if the separate process window crashes, the main window can detect this and take appropriate action, so not the whole app is crashing.

only if we could have the handle returned from the new window, so we can keep control of it.

@rogerwang please make sure the fix is in the next build, as I saw it was already fixed in the chromium upstream?

This will allow us to build great speedy apps with nwjs with great reliability!

**Comment by gpetrov** [comment]
@rogerwang - were you able to address this issue in 0.18.7? 

We really need urgently the handle to the new window running in separate process so we can communicate with it more reliable.

**Comment by rogerwang** [comment]
@gpetrov unfortunately this is not to be fixed in 0.18 as it require more important changes than expected.

To work around it before that and offload heavy duty tasks, you may consider using Node in Web workers: http://nwjs.io/blog/v0.18.4/

**Comment by gpetrov** [comment]
That is a sad news - will it be in the 0.19?

We tried the WebWorkers but unfortunately they don't work for us first then don't support WebRTC yet - and second and some of the modules we use, use process.binding('...') and that gave us errors in webWorkers

**Comment by alexsmith540** [comment]
I have a workaround but not sure if you're going to like it...
So I have a multi screen app and each screen needs its own renderer process hence the need for new_instance. You can use inject_js_start to inject websocket client code/libs into the windows. If you set an id when spawning the window it's present inside nw.Window.get().appWindow.id (or something like that sorry at an airport on mobile, ha). I then send that ID back to the nodejs socket server and boom: communication. HackyAF but we absolutely needed features from 0.18 and couldn't stay on 0.12 anymore...

**Comment by rogerwang** [comment]
@gpetrov I'm afraid I can't promise a deadline but I'll work on this issue at the highest priority.

**Comment by kailniris** [comment]
I don't really understand this.
new_instance means new render process which means new thread, new memory heap. JavaScript is designed to be single threaded , you can't access other threads memory heap so you can't just access other new_instance windows variable that easily.
Am I missing something important here?

**Comment by gpetrov** [comment]
thanks @rogerwang ! really appreciate it. 

Unfortunately we hit another road block in webRTC ... seems even if loaded in separate process, due to some chromium bugs in RTCPeerConnection.close it hangs not only the child process web rendered but also the main web renderer! And we thought those will be totally separated when ran with new_instance
See also https://github.com/feross/webtorrent/issues/903

@alexsmith540 we got the communication working with https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API which works great. It is only that we can't kill the client process if it gets stuck.

**Comment by rogerwang** [comment]
Please submit an issue for webrtc bug. Thanks.

**Comment by rogerwang** [comment]
And please submit issues you found in Node + Web worker.

**Comment by rogerwang** [comment]
reassign to @ghostoy since we found another way to support this.

**Comment by semmel** [comment]
This issue is still a major roadblock for us 🚧 🚧

We cannot load _remote_ pages in popup Windows using `nw.Window.open` unless we grant them _full local Node rights_ (ie. make them Node Frames by including their urls in the `node-remote`manifest entry). 

Because unless they are Node Frames the `nwWindow` argument in the callback to `nw.Window.open` will be `undefined`.

But that's a major security concern: Granting full _local execution rights_ to _remote source code_. 
Since we don't want to do that we have to remain on `nwjs 0.16.1` 😢

**Comment by rogerwang** [comment]
@semmel before we fix it please see the workaound in the comment earlier: https://github.com/nwjs/nw.js/issues/4418#issuecomment-260931505
Another is wrap the remote page in the new window in a webview and write your code outside the webview.
The communication between windows in different processes can be done via [BroadcastChannel](https://developer.mozilla.org/en-US/docs/Web/API/BroadcastChannel)

**Comment by rogerwang** [comment]
btw, if you use webview in the new window to load external page, `new_instance` is not needed at all, because pages in webview are guaranteed to be loaded in a new renderer process.

**Comment by semmel** [comment]
@rogerwang Thank you very much for the comments!
I was able to discover yet another workaround:
In `package.json` when I set `"main": "local.html", ` then the callback in 
```javascript
// local.html
nw.Window.open("http://remote/page.html", {}, function callback(nwWindow){
   assert(nwWindow !== undefined) // works fine
});
```
gets delivered the right `nwWindow` object, also with remote/page **not** being a Node Frame. (no bug) ✅ 

However when in `package.json`  I set `"main": "app.js", ` , and then
```javascript
// app.js
//.. some preparations
nw.Window.open("local.html", {}, function callback(nwWindow){
   assert(nwWindow !== undefined) // works fine
});
```
but now
```javascript
// local.html
nw.Window.open("http://remote/page.html", {}, function callback(nwWindow){
   assert(nwWindow !== undefined) // FAILS!
});
```
🐛 
So I refactored my app to launch the main html page `local.html`directly from the manifest file on start-up. So far it looks good. :crossed_fingers:

**Comment by mscreenie** [comment]
Still an issue in 21? Kindly see if this can be fixed.

**Comment by Christywl** [comment]
Reproduce on Linux/Windows with nwjs-sdk-v0.21.6. When opening a new window with "new_instance" set to true, the Window.open callback returns an undefined object

**Comment by vankasteelj** [comment]
isn't that normal? "new instance" litteraly means opening a brand new instance, hence separate from the main thread. What's worrying is that the callback is unefined if the file or remote url isnt working, for example try to open www.nonsense-gngngngn.com in a new window

**Comment by mscreenie** [comment]
Any update on the callback?

**Comment by gomestai** [comment]
so are there any other way to pass variable?

**Comment by anesuc** [comment]
I am working on an OS based on node, i.e using nw js for Apps. Right now I have it running Apps as child processes due to this very bug. And as a result things like the browser (which uses webview) freezes the whole thing at times because they are running off the same process. Essentially I have to halt third party Apps from being used until this problem is resolved, which is sad,

**Comment by kailniris** [comment]
@anesuc Why don't you open third party apps in new instances?

**Comment by anesuc** [comment]
@kailniris The main system instance needs full access to the Apps so that I can control them which also include things such as theming etc

**Comment by Rhodolite** [comment]
This is an issue for me still.

In particular with the following:

 nw.Window.get().showDevTools(
                null,
                function(dev_win){
                    console.log('dev_win:', dev_win);
                    console.log('arguments:', arguments.length);
                }
            );

This will log an undefined with 0 arguments (it will also, properly open up the Developer Tools window)

From the documentation at http://docs.nwjs.io/en/latest/References/Window/#winshowdevtoolsiframe-callback

"callback(dev_win) {Function} callback with the native window of the DevTools window."

I'm reporting this under this bug since as far as I can tell the two are related & the same issue

**Comment by f3l1x** [comment]
@rogerwang Hi, is there any progress here? Can I help you with something?

It would be nice to have some working workaround. :-) Will Broadcast/Channel API work?

**Comment by rogerwang** [comment]
Hi Broadcast Channel would totally work. Using it to bridge the JS part of NW API implementation in different windows is the plan. ghostoy was working on it but he left the team later and the work has been stalled. Feel free to pick it up if you like

**Comment by f3l1x** [comment]
Thanks @rogerwang for the quick answer. Is there any way how to pass any args or something to the new instance?

```
nw.Window.open('https://github.com', {new_instance: true, args: {foo: bar}});
```

**Comment by lilliesAndRoses** [comment]
It looks like the new window does get created.
But the challenge is getting its handle. APIs like chrome.app.window.getAll also seem to be failing. 
Is there any way to get that handle? 
How can two windows created in separate renderer process communicate?

**Comment by rogerwang** [comment]
At least BroadcastChannel API can be used to communicate between windows.

**Comment by rogerwang** [comment]
This would be fixed after #5875, which has some progress recently.

**Comment by anesuc** [comment]
This is a great development!

**Comment by tolrtd** [comment]
Will this fix #6724 as well? That one is fairly urgent for me.

**Comment by Adrriver** [comment]
Almost two years later, this is defect is still without resolution??

**Comment by rogerwang** [comment]
The depending issue #5875 is under active development and will be released soon. We'll need testers for that issue. Once it's resolved this one will be fixed easily. Thanks.

**Comment by rogerwang** [comment]
0.35.4 was just released with a testing version of #5875. This issue is fixed under it. Please see more information in that issue. Thanks.

**Comment by tolrtd** [comment]
Thank you Roger for this project. You are enabling many people to create the future!

**Comment by eiro10** [comment]
Still happening in 0.40.1-sdk-x64
